### PR TITLE
fix pca test for cov est (WIP)

### LINKED ISSUE #2174: BUG: Covariance calculation with mode='pca'
I modified the test in `test_cov.py` to test all covareance estimation methods around here:

https://github.com/mne-tools/mne-python/blob/master/mne/tests/test_cov.py#L444

This effectively added a test for the `pca` mode, which didn't seem to be tested before, and it failed. cc @dengemann

### PR DESCRIPTION
**Pull request by agramfort**
closes #2174

ping @dengemann

### CODE REVIEW

File: mne/tests/test_cov.py [hunk 1]
```diff
@@ -435,20 +435,17 @@ def test_compute_covariance_auto_reg():
     events_merged = merge_events(events, event_ids, 1234)
     picks = pick_types(raw.info, meg='mag', eeg=False)
     epochs = Epochs(raw, events_merged, 1234, tmin=-0.2, tmax=0,
-                    picks=picks[:5], baseline=(-0.2, -0.1), proj=True,
+                    picks=picks[:7], baseline=(-0.2, -0.1), proj=True,
```
Comment (agramfort) [inline mne/tests/test_cov.py]: I had to change to 7 for numerical issues.

File: mne/tests/test_cov.py [hunk 2]
```diff
@@ -435,20 +435,17 @@ def test_compute_covariance_auto_reg():
...
                     reject=reject, preload=True)
     epochs.crop(None, 0)[:10]
 
-    method_params = dict(factor_analysis=dict(iter_n_components=[30]),
-                         pca=dict(iter_n_components=[30]))
+    method_params = dict(factor_analysis=dict(iter_n_components=[3]),
+                         pca=dict(iter_n_components=[3]))
```
Comment (agramfort) [inline mne/tests/test_cov.py]: using more components than number of picks does not make sense. That was the original issue.

File: mne/tests/test_cov.py [hunk 3]
```diff
@@ -435,20 +435,17 @@ def test_compute_covariance_auto_reg():
...
-        covs = compute_covariance(epochs, method='auto',
-                                  method_params=method_params,
-                                  projs=True,
-                                  return_estimators=True)
-        warnings.simplefilter('always')
-        assert_equal(len(w), 1)
+    covs = compute_covariance(epochs, method='auto',
+                              method_params=method_params,
+                              projs=True,
+                              return_estimators=True)
```
Comment (agramfort) [inline mne/tests/test_cov.py]: I have no warning anymore.

File: mne/tests/test_cov.py [hunk 4]
```diff
@@ -456,21 +453,20 @@ def test_compute_covariance_auto_reg():
...
     assert_equal(set([c['method'] for c in cov3]),
                  set(methods))
 
     # projs not allowed with FA or PCA
-    assert_raises(ValueError, compute_covariance, epochs, method='pca',
-                  projs=True)
+    # XXX : projs does not seem to be allowed to be a bool !!!
+    # and no error is raised.
+    # assert_raises(ValueError, compute_covariance, epochs, method='pca',
+    #               projs=True)
```
Comment (agramfort) [inline mne/tests/test_cov.py]: @dengemann this is weird as the code behaves as it projs was always a list of proj vectors but you pass bool in the tests. There is also to ValueError for me here.

feel free to take over...

Comment (dengemann) [inline mne/tests/test_cov.py]: Taking a look later, thanks!

2015-06-13 16:17 GMT+02:00 Alexandre Gramfort <notifications@github.com>:

**Comment by dengemann** [comment]
@agramfort pushed a solution. We can merge if Travis is happy too.

**Pull request merged**